### PR TITLE
Paypal auto capture

### LINKED ISSUE #10563: Paypal - 'Capture' payment when status changed
With Paypal Standard (the built in Paypal gateway), when 'Payment Action' is set to 'Authorize' - changing the order status to 'complete' or 'processing'  does not result in the payment being captured.

Looks like this could be possible with the [Paypal API](https://developer.paypal.com/docs/integration/direct/capture-payment/), and would be consistent with the Stripe gateway behaviour - capture_payment();

Logged on ideas board:
[http://ideas.woothemes.com/forums/133476-woocommerce/suggestions/3945761--pre-authorize-only-checkout-for-paypal-and-stripe
](http://ideas.woothemes.com/forums/133476-woocommerce/suggestions/3945761--pre-authorize-only-checkout-for-paypal-and-stripe)

WC Version: 2.5.3
WP Version: 4.4.2

### PR DESCRIPTION
**Pull request by mikejolley**
Closes #10563

When using authorizations, captures funds automatically when the order status goes to processing or completed.

To test this, you’ll need a paypal sandbox account (business) and API creds.

### CODE REVIEW

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+
+	/**
+	 * Refund an order via PayPal.
+	 * @param  WC_Order $order
+	 * @param  float    $amount
+	 * @param  string   $reason
+	 * @param  bool     $sandbox
+	 * @return array|wp_error The parsed response from paypal, or a WP_Error object
```
Comment (claudiosmweb) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php]: `array|WP_Error`

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+		return $response_array;
+	}
+}
+
+/**
+ * Here for backwards compatibility.
+ * @since 2.7.0
+ */
+class WC_Gateway_Paypal_Refund extends WC_Gateway_Paypal_API_Handler {
```
Comment (claudiosmweb) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php]: Why 2 classes in the same file?

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	/** @var string API Username */
+	public static $api_username;
+
+	/** @var string API Password */
+	public static $api_password;
+
+	/** @var string API Signature */
+	public static $api_signature;
+
+	/** @var string API Signature */
```
Comment (justinshreve) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php]: Incorrect comment.

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+	public static $sandbox = false;
+
+	/**
+	 * Get capture request args.
+	 * See https://developer.paypal.com/docs/classic/api/merchant/DoCapture_API_Operation_NVP/.
+	 * @param  WC_Order $order
+	 * @param  float    $amount
+	 * @param  string   $reason
```
Comment (justinshreve) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php]: There is no `$reason` parameter.

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	 * @return object Either an object of name value pairs for a success, or a WP_ERROR object.
+	 */
+	public static function do_capture( $order, $amount = null ) {
+		$raw_response = wp_safe_remote_post(
+			self::$sandbox ? 'https://api-3t.sandbox.paypal.com/nvp' : 'https://api-3t.paypal.com/nvp',
+			array(
+				'method'      => 'POST',
+				'body'        => self::get_capture_request( $order, $amount ),
+				'timeout'     => 70,
+				'user-agent'  => 'WooCommerce',
```
Comment (justinshreve) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php]: We should include a version here like we do for other requests where we include a user-agent.

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	public static function get_refund_request( $order, $amount = null, $reason = '' ) {
+		$request = array(
+			'VERSION'       => '84.0',
+			'SIGNATURE'     => self::$api_signature,
+			'USER'          => self::$api_username,
+			'PWD'           => self::$api_password,
+			'METHOD'        => 'RefundTransaction',
+			'TRANSACTIONID' => $order->get_transaction_id(),
+			'NOTE'          => html_entity_decode( wc_trim_string( $reason, 255 ), ENT_NOQUOTES, 'UTF-8' ),
+			'REFUNDTYPE'    => 'Full'
```
Comment (justinshreve) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php]: missing trailing comma

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	 */
+	public static function refund_transaction( $order, $amount = null, $reason = '' ) {
+		$raw_response = wp_safe_remote_post(
+			self::$sandbox ? 'https://api-3t.sandbox.paypal.com/nvp' : 'https://api-3t.paypal.com/nvp',
+			array(
+				'method'      => 'POST',
+				'body'        => self::get_refund_request( $order, $amount, $reason ),
+				'timeout'     => 70,
+				'user-agent'  => 'WooCommerce',
+				'httpversion' => '1.1'
```
Comment (justinshreve) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php]: Trailing comma.

File: includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php (hunk 2 above)
Comment (mikejolley) [inline includes/gateways/paypal/includes/class-wc-gateway-paypal-api-handler.php]: BW compat code. Didn't seem worth moving.

**Pull request merged**